protocol/x/clob/e2e/withdrawal_gating_test.go

```diff
@@ -0,0 +1,231 @@
+				testapp.UpdateGenesisDocWithAppStateForModule(
+					&genesis,
+					func(genesisState *satypes.GenesisState) {
+						genesisState.Subaccounts = tc.subaccounts
+					},
+				)
+				testapp.UpdateGenesisDocWithAppStateForModule(
+					&genesis,
+					func(genesisState *clobtypes.GenesisState) {
+						genesisState.ClobPairs = tc.clobPairs
+						genesisState.LiquidationsConfig = tc.liquidationConfig
+						genesisState.EquityTierLimitConfig = clobtypes.EquityTierLimitConfiguration{}
+					},
+				)
+				testapp.UpdateGenesisDocWithAppStateForModule(
+					&genesis,
+					func(genesisState *feetiertypes.GenesisState) {
+						genesisState.Params = constants.PerpetualFeeParamsNoFee
+					},
+				)
+				return genesis
+			}).Build()
+
+			ctx := tApp.AdvanceToBlock(2, testapp.AdvanceToBlockOptions{})
+
+			// Create all existing orders.
+			existingOrderMsgs := make([]clobtypes.MsgPlaceOrder, len(tc.placedMatchableOrders))
+			for i, matchableOrder := range tc.placedMatchableOrders {
+				existingOrderMsgs[i] = clobtypes.MsgPlaceOrder{Order: matchableOrder.MustGetOrder()}
+			}
+			for _, checkTx := range testapp.MustMakeCheckTxsWithClobMsg(ctx, tApp.App, existingOrderMsgs...) {
+				resp := tApp.CheckTx(checkTx)
+				require.Conditionf(t, resp.IsOK, "Expected CheckTx to succeed. Response: %+v", resp)
+			}
+
+			_, err := tApp.App.Server.LiquidateSubaccounts(ctx, &api.LiquidateSubaccountsRequest{
+				LiquidatableSubaccountIds:  tc.liquidatableSubaccountIds,
+				NegativeTncSubaccountIds:   tc.negativeTncSubaccountIds,
+				SubaccountOpenPositionInfo: clobtest.GetOpenPositionsFromSubaccounts(tc.subaccounts),
+			})
+			require.NoError(t, err)
+
+			// Verify test expectations.
+			ctx = tApp.AdvanceToBlock(4, testapp.AdvanceToBlockOptions{})
+			for _, expectedSubaccount := range tc.expectedSubaccounts {
+				require.Equal(
+					t,
+					expectedSubaccount,
+					tApp.App.SubaccountsKeeper.GetSubaccount(ctx, *expectedSubaccount.Id),
+				)
+			}
+			negativeTncSubaccountSeenAtBlock, exists := tApp.App.SubaccountsKeeper.GetNegativeTncSubaccountSeenAtBlock(ctx)
+			require.Equal(t, tc.expectedWithdrawalsGated, exists)
+			require.Equal(t, tc.expectedNegativeTncSubaccountSeenAtBlock, negativeTncSubaccountSeenAtBlock)
+
+			// Verify withdrawals are blocked by trying to create a transfer message that withdraws funds.
+			withdrawMsg := sendingtypes.MsgWithdrawFromSubaccount{
+				Sender:    constants.Carl_Num0,
+				Recipient: constants.Carl_Num0.Owner,
+				AssetId:   constants.Usdc.Id,
+				Quantums:  1,
+			}
+			for _, checkTx := range testapp.MustMakeCheckTxsWithSdkMsg(
+				ctx,
+				tApp.App,
+				testapp.MustMakeCheckTxOptions{
+					AccAddressForSigning: constants.Carl_Num0.Owner,
+					Gas:                  1000000,
+					FeeAmt:               constants.TestFeeCoins_5Cents,
+				},
+				&withdrawMsg,
+			) {
+				resp := tApp.CheckTx(checkTx)
+				require.Conditionf(t, resp.IsOK, "Expected CheckTx to succeed. Response: %+v", resp)
```

Comment:
Is there a reason why we can't reject withdrawals / transfers here? is it because we don't actually process the transfer until `DeliverTx`?

Reply:
yeah iirc txs only execute their handler in `DeliverTx` right? https://github.com/dydxprotocol/cosmos-sdk/blob/b48250d4e0efacbd40f3e0b980c1e1004d29e351/baseapp/baseapp.go#L1106-L1119

Reply:
hmm yeah not sure if it's worth building an ante handler for this 🤔 nbd for now

Reply:
yeah i dont think it is personally since this is an edge case

ideally frontend's query whether withdrawals are blocked before submitting in vast majority of cases